piptools/build.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+import collections
+import contextlib
+import os
+import pathlib
+import tempfile
+from dataclasses import dataclass
+from email.message import Message
+from importlib import metadata as importlib_metadata
+from typing import Any, Iterator
+
+import build
+import build.env
+import pyproject_hooks
+from pip._internal.req import InstallRequirement
+from pip._internal.req.constructors import install_req_from_line, parse_req_from_line
+
+PYPROJECT_TOML = "pyproject.toml"
+
+
+@dataclass
+class ProjectMetadata:
+    extras: tuple[str, ...]
+    requirements: tuple[InstallRequirement, ...]
+    build_requirements: tuple[InstallRequirement, ...]
+
+
+def build_project_metadata(
+    src_file: str,
+    build_distributions: tuple[str, ...],
+    *,
+    isolated: bool,
+    quiet: bool,
+) -> ProjectMetadata:
+    """
+    Return the metadata for a project.
+
+    Uses the ``prepare_metadata_for_build_wheel`` hook for the wheel metadata
+    if available, otherwise ``build_wheel``.
+
+    Uses the ``prepare_metadata_for_build_{dist}`` hook for each ``build_distributions``
+    if available.
+
+    :param src_file: Project source file
+    :param build_distributions: A tuple of build distributions to get the dependencies
+                                of (``sdist`` or ``wheel`` or ``editable``).
+    :param isolated: Whether to run invoke the backend in the current
+                     environment or to create an isolated one and invoke it
+                     there.
+    :param quiet: Whether to suppress the output of subprocesses.
+    """
+
+    src_dir = os.path.dirname(os.path.abspath(src_file))
+    with _create_project_builder(src_dir, isolated=isolated, quiet=quiet) as builder:
+        metadata = _build_project_wheel_metadata(builder)
+        extras = tuple(metadata.get_all("Provides-Extra") or ())
+        requirements = tuple(
+            _prepare_requirements(metadata=metadata, src_file=src_file)
+        )
+        build_requirements = tuple(
+            _prepare_build_requirements(
+                builder=builder,
+                src_file=src_file,
+                distributions=build_distributions,
+                package_name=_get_name(metadata),
+            )
+        )
+        return ProjectMetadata(
+            extras=extras,
+            requirements=requirements,
+            build_requirements=build_requirements,
+        )
+
+
+@contextlib.contextmanager
+def _create_project_builder(
+    src_dir: str, *, isolated: bool, quiet: bool
+) -> Iterator[build.ProjectBuilder]:
+    builder = build.ProjectBuilder(
+        os.fspath(src_dir),
+        runner=pyproject_hooks.quiet_subprocess_runner
+        if quiet
+        else pyproject_hooks.default_subprocess_runner,
```

Comment:
I've noticed that this PR makes `pip-compile` noisy by default, see details below.

<details>

<summary>Details</summary>

```toml
# pyproject.toml
[project]
name = "tmp"
version = "0.1"
dependencies = ["six"]

[build-system]
requires = ["setuptools"]
build-backend = "setuptools.build_meta"
```

```console
$ pip-compile pyproject.toml
running egg_info
writing tmp.egg-info/PKG-INFO
writing dependency_links to tmp.egg-info/dependency_links.txt
writing requirements to tmp.egg-info/requires.txt
writing top-level names to tmp.egg-info/top_level.txt
reading manifest file 'tmp.egg-info/SOURCES.txt'
writing manifest file 'tmp.egg-info/SOURCES.txt'
running dist_info
creating /var/folders/l0/lnq1ghps5yqc4vkgszlcz92m0000gp/T/tmp3roeb9hk/tmp.egg-info
writing /var/folders/l0/lnq1ghps5yqc4vkgszlcz92m0000gp/T/tmp3roeb9hk/tmp.egg-info/PKG-INFO
writing dependency_links to /var/folders/l0/lnq1ghps5yqc4vkgszlcz92m0000gp/T/tmp3roeb9hk/tmp.egg-info/dependency_links.txt
writing requirements to /var/folders/l0/lnq1ghps5yqc4vkgszlcz92m0000gp/T/tmp3roeb9hk/tmp.egg-info/requires.txt
writing top-level names to /var/folders/l0/lnq1ghps5yqc4vkgszlcz92m0000gp/T/tmp3roeb9hk/tmp.egg-info/top_level.txt
writing manifest file '/var/folders/l0/lnq1ghps5yqc4vkgszlcz92m0000gp/T/tmp3roeb9hk/tmp.egg-info/SOURCES.txt'
reading manifest file '/var/folders/l0/lnq1ghps5yqc4vkgszlcz92m0000gp/T/tmp3roeb9hk/tmp.egg-info/SOURCES.txt'
writing manifest file '/var/folders/l0/lnq1ghps5yqc4vkgszlcz92m0000gp/T/tmp3roeb9hk/tmp.egg-info/SOURCES.txt'
creating '/var/folders/l0/lnq1ghps5yqc4vkgszlcz92m0000gp/T/tmp3roeb9hk/tmp-0.1.dist-info'
#
# This file is autogenerated by pip-compile with Python 3.11
# by the following command:
#
#    pip-compile --strip-extras pyproject.toml
#
six==1.16.0
    # via tmp (pyproject.toml)
```

</details>